Automattic/jetpack
# WAF: Update reading values from targets

This PR fixes some bugs in how values were read from targets, and adds unit tests to cover all currently available targets.

It does go slightly beyond the original scope and makes some tweaks / bugfixes to other targets that could arguably be called unrelated. I felt they were small enough to go ahead and slip them into this PR (since I was also adding tests for them anyway), but if it makes sense to separate out the changes not related to using regex in target filters I can split those into another PR.

#### Changes proposed in this Pull Request:

* Expand `Waf_Request` class to provide all information about request (so that it can be more easily mocked for testing)
* Update how `ARGS` and `FILES` targets are handled to bring them closer to how ModSecurity reads them (mostly convert deeply nested arrays into shallow key/value arrays with a key that looks like `myvar[nested][array]`)
* Fix `Waf_Runtime` error that was preventing regular expressions from being used to read values from targets (`ARGS:/^my/` should return all request fields that start with `my`, not a single request field named `/^my/`).
* Fix `Waf_Runtime` error that was returning incorrect value for `REQUEST_PROTOCOL` target.
* Add unit tests to cover all new `Waf_Request` methods.
* Add unit tests to cover all supported targets for WAF rules.

#### Other information:


#### Jetpack product discussion


1201069996155224-as-1202036802073208

#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






* Confirm that the existing WAF rules do not support using regex in target filters:
```
> curl 'https://your-website/wp-admin/admin-ajax.php?action=nf_ajax_submit' --header 'referer: ?test::'
0
```
* Using the Jetpack Beta plugin, install this branch (`fix/waf-rules-enable-regex-targets`) of the Jetpack plugin.
* Disable and re-enable the Firewall in Jetpack settings (in order to force the `jetpack-dev` plugin to download rules)
* Run the same `curl` request to confirm that the firewall now supports using regex in target filters:
```
> curl 'https://your-website/wp-admin/admin-ajax.php?action=nf_ajax_submit' --header 'referer: ?test::'
rule 99101000 - reason Potential Ninja Forms RCE
```
* Check out the new unit tests in the `waf` package.
